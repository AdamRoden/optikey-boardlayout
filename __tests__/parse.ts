import example from "../src/xml/example"
import xmlParser, { toXml } from "../src/xml/parse"

test("parser", () => {
  expect(xmlParser(example)).toMatchSnapshot("xml to js")
  expect(toXml(xmlParser(example))).toMatchSnapshot("js to xml")
})
