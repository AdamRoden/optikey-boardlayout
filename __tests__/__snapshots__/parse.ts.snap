// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser: js to xml 1`] = `
"<Keyboard>
  <Name>MMB2s Keyboard</Name>
  <Height>40</Height>
  <SymbolMargin>1</SymbolMargin>
  <BorderThickness>0,0,0,0</BorderThickness>
  <Grid>
    <Rows>7</Rows>
    <Cols>12</Cols>
  </Grid>
  <Keys>
    <TextKey>
      <Row>0</Row>
      <Col>8</Col>
      <Text>7</Text>
      <Label>7</Label>
    </TextKey>
    <TextKey>
      <Row>0</Row>
      <Col>9</Col>
      <Text>8</Text>
      <Label>8</Label>
    </TextKey>
    <TextKey>
      <Row>0</Row>
      <Col>10</Col>
      <Text>9</Text>
      <Label>9</Label>
    </TextKey>
    <TextKey>
      <Row>2</Row>
      <Col>1</Col>
      <Text>\\\\</Text>
      <Label>\\\\</Label>
    </TextKey>
    <TextKey>
      <Row>0</Row>
      <Col>0</Col>
      <Text>~</Text>
      <Label>~</Label>
    </TextKey>
    <TextKey>
      <Row>1</Row>
      <Col>2</Col>
      <Text>$</Text>
      <Label>$</Label>
    </TextKey>
    <TextKey>
      <Row>1</Row>
      <Col>1</Col>
      <Text>&</Text>
      <Label>&</Label>
    </TextKey>
    <TextKey>
      <Row>3</Row>
      <Col>1</Col>
      <Text>^</Text>
      <Label>^</Label>
    </TextKey>
    <TextKey>
      <Row>2</Row>
      <Col>0</Col>
      <Text>!</Text>
      <Label>!</Label>
    </TextKey>
    <TextKey>
      <Row>3</Row>
      <Col>2</Col>
      <Text>?</Text>
      <Label>?</Label>
    </TextKey>
    <TextKey>
      <Row>4</Row>
      <Col>10</Col>
      <Text>|</Text>
      <Label>|</Label>
    </TextKey>
    <TextKey>
      <Row>0</Row>
      <Col>5</Col>
      <Text>4</Text>
      <Label>4</Label>
    </TextKey>
    <TextKey>
      <Row>0</Row>
      <Col>6</Col>
      <Text>5</Text>
      <Label>5</Label>
    </TextKey>
    <TextKey>
      <Row>0</Row>
      <Col>7</Col>
      <Text>6</Text>
      <Label>6</Label>
    </TextKey>
    <TextKey>
      <Row>1</Row>
      <Col>3</Col>
      <Text>/</Text>
      <Label>/</Label>
    </TextKey>
    <TextKey>
      <Row>4</Row>
      <Col>0</Col>
      <Text>:</Text>
      <Label>:</Label>
    </TextKey>
    <TextKey>
      <Row>3</Row>
      <Col>6</Col>
      <Text>f</Text>
      <ShiftUpLabel>f</ShiftUpLabel>
      <ShiftDownLabel>F</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>3</Row>
      <Col>7</Col>
      <Text>g</Text>
      <ShiftUpLabel>g</ShiftUpLabel>
      <ShiftDownLabel>G</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>3</Row>
      <Col>8</Col>
      <Text>h</Text>
      <ShiftUpLabel>h</ShiftUpLabel>
      <ShiftDownLabel>H</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>2</Row>
      <Col>2</Col>
      <Text>q</Text>
      <ShiftUpLabel>q</ShiftUpLabel>
      <ShiftDownLabel>Q</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>3</Row>
      <Col>9</Col>
      <Text>j</Text>
      <ShiftUpLabel>j</ShiftUpLabel>
      <ShiftDownLabel>J</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>1</Row>
      <Col>8</Col>
      <Text>%</Text>
      <Label>%</Label>
    </TextKey>
    <TextKey>
      <Row>0</Row>
      <Col>2</Col>
      <Text>1</Text>
      <Label>1</Label>
    </TextKey>
    <TextKey>
      <Row>0</Row>
      <Col>3</Col>
      <Text>2</Text>
      <Label>2</Label>
    </TextKey>
    <TextKey>
      <Row>4</Row>
      <Col>9</Col>
      <Text>m</Text>
      <ShiftUpLabel>m</ShiftUpLabel>
      <ShiftDownLabel>M</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>4</Row>
      <Col>8</Col>
      <Text>n</Text>
      <ShiftUpLabel>n</ShiftUpLabel>
      <ShiftDownLabel>N</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>2</Row>
      <Col>11</Col>
      <Text>p</Text>
      <ShiftUpLabel>p</ShiftUpLabel>
      <ShiftDownLabel>P</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>4</Row>
      <Col>7</Col>
      <Text>b</Text>
      <ShiftUpLabel>b</ShiftUpLabel>
      <ShiftDownLabel>B</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>3</Row>
      <Col>10</Col>
      <Text>k</Text>
      <ShiftUpLabel>k</ShiftUpLabel>
      <ShiftDownLabel>K</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>4</Row>
      <Col>6</Col>
      <Text>v</Text>
      <ShiftUpLabel>v</ShiftUpLabel>
      <ShiftDownLabel>V</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>3</Row>
      <Col>0</Col>
      <Text>-</Text>
      <Label>-</Label>
    </TextKey>
    <TextKey>
      <Row>0</Row>
      <Col>4</Col>
      <Text>3</Text>
      <Label>3</Label>
    </TextKey>
    <TextKey>
      <Row>1</Row>
      <Col>6</Col>
      <Text>_</Text>
      <Label>_</Label>
    </TextKey>
    <TextKey>
      <Row>6</Row>
      <Col>2</Col>
      <Text>;</Text>
      <Label>;</Label>
    </TextKey>
    <TextKey>
      <Row>1</Row>
      <Col>0</Col>
      <Text>@</Text>
      <Label>@</Label>
    </TextKey>
    <TextKey>
      <Row>1</Row>
      <Col>7</Col>
      <Text>\\"</Text>
      <Label>\\"</Label>
    </TextKey>
    <TextKey>
      <Row>1</Row>
      <Col>5</Col>
      <Text>'</Text>
      <Label>'</Label>
    </TextKey>
    <TextKey>
      <Row>4</Row>
      <Col>11</Col>
      <Text>#</Text>
      <Label>#</Label>
    </TextKey>
    <TextKey>
      <Row>0</Row>
      <Col>1</Col>
      <Text>0</Text>
      <Label>0</Label>
    </TextKey>
    <TextKey>
      <Row>3</Row>
      <Col>11</Col>
      <Text>l</Text>
      <ShiftUpLabel>l</ShiftUpLabel>
      <ShiftDownLabel>L</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>3</Row>
      <Col>4</Col>
      <Text>s</Text>
      <ShiftUpLabel>s</ShiftUpLabel>
      <ShiftDownLabel>S</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>2</Row>
      <Col>5</Col>
      <Text>r</Text>
      <ShiftUpLabel>r</ShiftUpLabel>
      <ShiftDownLabel>R</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>2</Row>
      <Col>6</Col>
      <Text>t</Text>
      <ShiftUpLabel>t</ShiftUpLabel>
      <ShiftDownLabel>T</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>4</Row>
      <Col>5</Col>
      <Text>c</Text>
      <ShiftUpLabel>c</ShiftUpLabel>
      <ShiftDownLabel>C</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>3</Row>
      <Col>5</Col>
      <Text>d</Text>
      <ShiftUpLabel>d</ShiftUpLabel>
      <ShiftDownLabel>D</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>2</Row>
      <Col>3</Col>
      <Text>w</Text>
      <ShiftUpLabel>w</ShiftUpLabel>
      <ShiftDownLabel>W</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>1</Row>
      <Col>4</Col>
      <Text>+</Text>
      <Label>+</Label>
    </TextKey>
    <TextKey>
      <Row>4</Row>
      <Col>1</Col>
      <Text>&lt;</Text>
      <Label>&lt;</Label>
    </TextKey>
    <TextKey>
      <Row>4</Row>
      <Col>2</Col>
      <Text>=</Text>
      <Label>=</Label>
    </TextKey>
    <TextKey>
      <Row>3</Row>
      <Col>3</Col>
      <Text>a</Text>
      <ShiftUpLabel>a</ShiftUpLabel>
      <ShiftDownLabel>A</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>2</Row>
      <Col>4</Col>
      <Text>e</Text>
      <ShiftUpLabel>e</ShiftUpLabel>
      <ShiftDownLabel>E</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>2</Row>
      <Col>9</Col>
      <Text>i</Text>
      <ShiftUpLabel>i</ShiftUpLabel>
      <ShiftDownLabel>I</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>2</Row>
      <Col>10</Col>
      <Text>o</Text>
      <ShiftUpLabel>o</ShiftUpLabel>
      <ShiftDownLabel>O</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>2</Row>
      <Col>8</Col>
      <Text>u</Text>
      <ShiftUpLabel>u</ShiftUpLabel>
      <ShiftDownLabel>U</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>4</Row>
      <Col>4</Col>
      <Text>x</Text>
      <ShiftUpLabel>x</ShiftUpLabel>
      <ShiftDownLabel>X</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>4</Row>
      <Col>3</Col>
      <Text>y</Text>
      <ShiftUpLabel>y</ShiftUpLabel>
      <ShiftDownLabel>Y</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>2</Row>
      <Col>7</Col>
      <Text>z</Text>
      <ShiftUpLabel>z</ShiftUpLabel>
      <ShiftDownLabel>Z</ShiftDownLabel>
    </TextKey>
    <TextKey>
      <Row>5</Row>
      <Col>1</Col>
      <Text>&gt;</Text>
      <Label>&gt;</Label>
    </TextKey>
    <TextKey>
      <Row>5</Row>
      <Col>2</Col>
      <Text>,</Text>
      <Label>,</Label>
    </TextKey>
    <TextKey>
      <Row>5</Row>
      <Col>3</Col>
      <Text>.</Text>
      <Label>.</Label>
    </TextKey>
    <TextKey>
      <Row>5</Row>
      <Col>4</Col>
      <Text>(</Text>
      <Label>(</Label>
    </TextKey>
    <TextKey>
      <Row>5</Row>
      <Col>5</Col>
      <Text>)</Text>
      <Label>)</Label>
    </TextKey>
    <TextKey>
      <Row>5</Row>
      <Col>6</Col>
      <Text>[</Text>
      <Label>[</Label>
    </TextKey>
    <TextKey>
      <Row>5</Row>
      <Col>7</Col>
      <Text>]</Text>
      <Label>]</Label>
    </TextKey>
    <TextKey>
      <Row>5</Row>
      <Col>8</Col>
      <Text>{</Text>
      <Label>{</Label>
    </TextKey>
    <TextKey>
      <Row>5</Row>
      <Col>9</Col>
      <Text>}</Text>
      <Label>}</Label>
    </TextKey>
    <TextKey>
      <Row>6</Row>
      <Col>4</Col>
      <Symbol>TabIcon</Symbol>
      <Text>&#9;</Text>
    </TextKey>
    <TextKey>
      <Row>6</Row>
      <Col>5</Col>
      <Width>2</Width>
      <Symbol>SpaceIcon</Symbol>
      <Text>&#32;</Text>
    </TextKey>
    <TextKey>
      <Row>6</Row>
      <Col>7</Col>
      <Symbol>EnterIcon</Symbol>
      <Text>&#13;</Text>
    </TextKey>
    <ActionKey>
      <Row>0</Row>
      <Col>11</Col>
      <Symbol>MenuIcon</Symbol>
      <Action>MenuKeyboard</Action>
    </ActionKey>
    <ActionKey>
      <Row>6</Row>
      <Col>8</Col>
      <Symbol>BackOneIcon</Symbol>
      <Action>BackOne</Action>
    </ActionKey>
    <ActionKey>
      <Row>5</Row>
      <Col>0</Col>
      <Label>Win</Label>
      <Action>LeftWin</Action>
    </ActionKey>
    <ActionKey>
      <Row>5</Row>
      <Col>10</Col>
      <Symbol>UpArrowKeyIcon</Symbol>
      <Action>ArrowUp</Action>
    </ActionKey>
    <ActionKey>
      <Row>5</Row>
      <Col>11</Col>
      <Label>Esc</Label>
      <Action>Escape</Action>
    </ActionKey>
    <ActionKey>
      <Row>6</Row>
      <Col>0</Col>
      <Label>Ctrl</Label>
      <Action>LeftCtrl</Action>
    </ActionKey>
    <ActionKey>
      <Row>6</Row>
      <Col>1</Col>
      <Label>Alt</Label>
      <Action>LeftAlt</Action>
    </ActionKey>
    <ActionKey>
      <Row>6</Row>
      <Col>3</Col>
      <Symbol>ShiftIcon</Symbol>
      <Action>LeftShift</Action>
    </ActionKey>
    <ActionKey>
      <Row>6</Row>
      <Col>9</Col>
      <Symbol>LeftArrowKeyIcon</Symbol>
      <Action>ArrowLeft</Action>
    </ActionKey>
    <ActionKey>
      <Row>6</Row>
      <Col>10</Col>
      <Symbol>DownArrowKeyIcon</Symbol>
      <Action>ArrowDown</Action>
    </ActionKey>
    <ActionKey>
      <Row>6</Row>
      <Col>11</Col>
      <Symbol>RightArrowKeyIcon</Symbol>
      <Action>ArrowRight</Action>
    </ActionKey>
  </Keys>
</Keyboard>"
`;

exports[`parser: xml to js 1`] = `
Object {
  "Keyboard": Object {
    "BorderThickness": "0,0,0,0",
    "Grid": Object {
      "Cols": 12,
      "Rows": 7,
    },
    "Height": 40,
    "Keys": Array [
      Object {
        "Col": 8,
        "Label": "7",
        "Row": 0,
        "Text": "7",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 9,
        "Label": "8",
        "Row": 0,
        "Text": "8",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 10,
        "Label": "9",
        "Row": 0,
        "Text": "9",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 1,
        "Label": "\\\\",
        "Row": 2,
        "Text": "\\\\",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 0,
        "Label": "~",
        "Row": 0,
        "Text": "~",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 2,
        "Label": "$",
        "Row": 1,
        "Text": "$",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 1,
        "Label": "&",
        "Row": 1,
        "Text": "&",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 1,
        "Label": "^",
        "Row": 3,
        "Text": "^",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 0,
        "Label": "!",
        "Row": 2,
        "Text": "!",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 2,
        "Label": "?",
        "Row": 3,
        "Text": "?",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 10,
        "Label": "|",
        "Row": 4,
        "Text": "|",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Action": "MenuKeyboard",
        "Col": 11,
        "Row": 0,
        "Symbol": "MenuIcon",
        "type": "ActionKey",
        "useSymbol": true,
      },
      Object {
        "Col": 5,
        "Label": "4",
        "Row": 0,
        "Text": "4",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 6,
        "Label": "5",
        "Row": 0,
        "Text": "5",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 7,
        "Label": "6",
        "Row": 0,
        "Text": "6",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 3,
        "Label": "/",
        "Row": 1,
        "Text": "/",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 0,
        "Label": ":",
        "Row": 4,
        "Text": ":",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 6,
        "Row": 3,
        "ShiftDownLabel": "F",
        "ShiftUpLabel": "f",
        "Text": "f",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 7,
        "Row": 3,
        "ShiftDownLabel": "G",
        "ShiftUpLabel": "g",
        "Text": "g",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 8,
        "Row": 3,
        "ShiftDownLabel": "H",
        "ShiftUpLabel": "h",
        "Text": "h",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 2,
        "Row": 2,
        "ShiftDownLabel": "Q",
        "ShiftUpLabel": "q",
        "Text": "q",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 9,
        "Row": 3,
        "ShiftDownLabel": "J",
        "ShiftUpLabel": "j",
        "Text": "j",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 8,
        "Label": "%",
        "Row": 1,
        "Text": "%",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Action": "BackOne",
        "Col": 8,
        "Row": 6,
        "Symbol": "BackOneIcon",
        "type": "ActionKey",
        "useSymbol": true,
      },
      Object {
        "Col": 2,
        "Label": "1",
        "Row": 0,
        "Text": "1",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 3,
        "Label": "2",
        "Row": 0,
        "Text": "2",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 9,
        "Row": 4,
        "ShiftDownLabel": "M",
        "ShiftUpLabel": "m",
        "Text": "m",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 8,
        "Row": 4,
        "ShiftDownLabel": "N",
        "ShiftUpLabel": "n",
        "Text": "n",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 11,
        "Row": 2,
        "ShiftDownLabel": "P",
        "ShiftUpLabel": "p",
        "Text": "p",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 7,
        "Row": 4,
        "ShiftDownLabel": "B",
        "ShiftUpLabel": "b",
        "Text": "b",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 10,
        "Row": 3,
        "ShiftDownLabel": "K",
        "ShiftUpLabel": "k",
        "Text": "k",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 6,
        "Row": 4,
        "ShiftDownLabel": "V",
        "ShiftUpLabel": "v",
        "Text": "v",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 0,
        "Label": "-",
        "Row": 3,
        "Text": "-",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 4,
        "Label": "3",
        "Row": 0,
        "Text": "3",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 6,
        "Label": "_",
        "Row": 1,
        "Text": "_",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 2,
        "Label": ";",
        "Row": 6,
        "Text": ";",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 0,
        "Label": "@",
        "Row": 1,
        "Text": "@",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 7,
        "Label": "\\"",
        "Row": 1,
        "Text": "\\"",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 5,
        "Label": "'",
        "Row": 1,
        "Text": "'",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 11,
        "Label": "#",
        "Row": 4,
        "Text": "#",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 1,
        "Label": "0",
        "Row": 0,
        "Text": "0",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 11,
        "Row": 3,
        "ShiftDownLabel": "L",
        "ShiftUpLabel": "l",
        "Text": "l",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 4,
        "Row": 3,
        "ShiftDownLabel": "S",
        "ShiftUpLabel": "s",
        "Text": "s",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 5,
        "Row": 2,
        "ShiftDownLabel": "R",
        "ShiftUpLabel": "r",
        "Text": "r",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 6,
        "Row": 2,
        "ShiftDownLabel": "T",
        "ShiftUpLabel": "t",
        "Text": "t",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 5,
        "Row": 4,
        "ShiftDownLabel": "C",
        "ShiftUpLabel": "c",
        "Text": "c",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 5,
        "Row": 3,
        "ShiftDownLabel": "D",
        "ShiftUpLabel": "d",
        "Text": "d",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 3,
        "Row": 2,
        "ShiftDownLabel": "W",
        "ShiftUpLabel": "w",
        "Text": "w",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 4,
        "Label": "+",
        "Row": 1,
        "Text": "+",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 1,
        "Label": "<",
        "Row": 4,
        "Text": "<",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 2,
        "Label": "=",
        "Row": 4,
        "Text": "=",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 3,
        "Row": 3,
        "ShiftDownLabel": "A",
        "ShiftUpLabel": "a",
        "Text": "a",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 4,
        "Row": 2,
        "ShiftDownLabel": "E",
        "ShiftUpLabel": "e",
        "Text": "e",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 9,
        "Row": 2,
        "ShiftDownLabel": "I",
        "ShiftUpLabel": "i",
        "Text": "i",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 10,
        "Row": 2,
        "ShiftDownLabel": "O",
        "ShiftUpLabel": "o",
        "Text": "o",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 8,
        "Row": 2,
        "ShiftDownLabel": "U",
        "ShiftUpLabel": "u",
        "Text": "u",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 4,
        "Row": 4,
        "ShiftDownLabel": "X",
        "ShiftUpLabel": "x",
        "Text": "x",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 3,
        "Row": 4,
        "ShiftDownLabel": "Y",
        "ShiftUpLabel": "y",
        "Text": "y",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Col": 7,
        "Row": 2,
        "ShiftDownLabel": "Z",
        "ShiftUpLabel": "z",
        "Text": "z",
        "caseSensitive": true,
        "type": "TextKey",
      },
      Object {
        "Action": "LeftWin",
        "Col": 0,
        "Label": "Win",
        "Row": 5,
        "caseSensitive": false,
        "type": "ActionKey",
        "useSymbol": false,
      },
      Object {
        "Col": 1,
        "Label": ">",
        "Row": 5,
        "Text": ">",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 2,
        "Label": ",",
        "Row": 5,
        "Text": ",",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 3,
        "Label": ".",
        "Row": 5,
        "Text": ".",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 4,
        "Label": "(",
        "Row": 5,
        "Text": "(",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 5,
        "Label": ")",
        "Row": 5,
        "Text": ")",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 6,
        "Label": "[",
        "Row": 5,
        "Text": "[",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 7,
        "Label": "]",
        "Row": 5,
        "Text": "]",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 8,
        "Label": "{",
        "Row": 5,
        "Text": "{",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Col": 9,
        "Label": "}",
        "Row": 5,
        "Text": "}",
        "caseSensitive": false,
        "type": "TextKey",
        "useSymbol": false,
      },
      Object {
        "Action": "ArrowUp",
        "Col": 10,
        "Row": 5,
        "Symbol": "UpArrowKeyIcon",
        "type": "ActionKey",
        "useSymbol": true,
      },
      Object {
        "Action": "Escape",
        "Col": 11,
        "Label": "Esc",
        "Row": 5,
        "caseSensitive": false,
        "type": "ActionKey",
        "useSymbol": false,
      },
      Object {
        "Action": "LeftCtrl",
        "Col": 0,
        "Label": "Ctrl",
        "Row": 6,
        "caseSensitive": false,
        "type": "ActionKey",
        "useSymbol": false,
      },
      Object {
        "Action": "LeftAlt",
        "Col": 1,
        "Label": "Alt",
        "Row": 6,
        "caseSensitive": false,
        "type": "ActionKey",
        "useSymbol": false,
      },
      Object {
        "Action": "LeftShift",
        "Col": 3,
        "Row": 6,
        "Symbol": "ShiftIcon",
        "type": "ActionKey",
        "useSymbol": true,
      },
      Object {
        "Col": 4,
        "Row": 6,
        "Symbol": "TabIcon",
        "Text": "&#9;",
        "type": "TextKey",
        "useSymbol": true,
      },
      Object {
        "Col": 5,
        "Row": 6,
        "Symbol": "SpaceIcon",
        "Text": "&#32;",
        "Width": "2",
        "type": "TextKey",
        "useSymbol": true,
      },
      Object {
        "Col": 7,
        "Row": 6,
        "Symbol": "EnterIcon",
        "Text": "&#13;",
        "type": "TextKey",
        "useSymbol": true,
      },
      Object {
        "Action": "ArrowLeft",
        "Col": 9,
        "Row": 6,
        "Symbol": "LeftArrowKeyIcon",
        "type": "ActionKey",
        "useSymbol": true,
      },
      Object {
        "Action": "ArrowDown",
        "Col": 10,
        "Row": 6,
        "Symbol": "DownArrowKeyIcon",
        "type": "ActionKey",
        "useSymbol": true,
      },
      Object {
        "Action": "ArrowRight",
        "Col": 11,
        "Row": 6,
        "Symbol": "RightArrowKeyIcon",
        "type": "ActionKey",
        "useSymbol": true,
      },
    ],
    "Name": "MMB2s Keyboard",
    "SymbolMargin": 1,
  },
}
`;
